Prediction clones of fedulya user from best models created in fedulya_classification.ipynb, averaging by all models, finding clones and quick review of user-clones.

In [1]:
from google.colab import drive
drive.mount('/content/drive')

Mounted at /content/drive


In [ ]:
import tensorflow as tf

In [ ]:
gpu_info = !nvidia-smi
gpu_info = '\n'.join(gpu_info)
if gpu_info.find('failed') >= 0:
  print('Not connected to a GPU')
else:
  print(gpu_info)

Wed Jan  5 00:40:16 2022       
+-----------------------------------------------------------------------------+
| NVIDIA-SMI 495.44       Driver Version: 460.32.03    CUDA Version: 11.2     |
|-------------------------------+----------------------+----------------------+
| GPU  Name        Persistence-M| Bus-Id        Disp.A | Volatile Uncorr. ECC |
| Fan  Temp  Perf  Pwr:Usage/Cap|         Memory-Usage | GPU-Util  Compute M. |
|                               |                      |               MIG M. |
|===============================+======================+======================|
|   0  Tesla P100-PCIE...  Off  | 00000000:00:04.0 Off |                    0 |
| N/A   34C    P0    26W / 250W |      0MiB / 16280MiB |      0%      Default |
|                               |                      |                  N/A |
+-------------------------------+----------------------+----------------------+
                                                                               
+-------

In [ ]:
from psutil import virtual_memory
ram_gb = virtual_memory().total / 1e9
print('Your runtime has {:.1f} gigabytes of available RAM\n'.format(ram_gb))

if ram_gb < 20:
  print('Not using a high-RAM runtime')
else:
  print('You are using a high-RAM runtime!')

Your runtime has 27.3 gigabytes of available RAM

You are using a high-RAM runtime!


In [3]:
import os
import time

import pandas as pd
import numpy as np

In [ ]:
!pip install ktrain

In [ ]:
import ktrain
from ktrain import text

In [20]:
Data = '/content/drive/MyDrive/Colab Notebooks/Projects/eva/Data/'

test_Messages_filename='all_fedulya_and_clones_messages.csv'
test_Messages_full_filename= os.path.join(Data, test_Messages_filename)

Authors_filename='Authors.csv'
Authors_full_filename= os.path.join(Data, Authors_filename)

CoreAuthors_filename='CoreAuthors.csv'
CoreAuthors_full_filename=os.path.join(Data, CoreAuthors_filename)

text_column='original_message'
classname='fedulya'
prediction_column='prediction'

Models = '/content/drive/MyDrive/Colab Notebooks/Projects/eva/Models/fedulya/'

threshold_fedulya_probability=0.9
threshold_frc_fedulya_messages=0.7

## Data for prediction

In [5]:
df=pd.read_csv(test_Messages_full_filename, error_bad_lines=False, index_col=False)

In [ ]:
len(df)

435070

In [ ]:
df.head()

Message_Id            Timestamp  ... prediction_9  prediction
0    68156918  2011-07-20 15:29:00  ...     0.317306    0.209526
1    65203211  2011-03-19 15:06:00  ...     0.000098    0.000103
2    72508255  2012-01-22 16:37:00  ...     0.000099    0.000103
3    81885636  2013-08-29 09:39:00  ...     0.000081    0.000090
4    81888444  2013-08-29 12:57:00  ...     0.000093    0.000115

[5 rows x 25 columns]

## Prediction clone or not

In [ ]:
prediction_column_lists=list()
ts_start = time.time()
for m in next(os.walk(Models))[1]:  
  ind= m.split('_')[len(m.split('_'))-1]
  prediction_column_lists.append('%s_%s'%(prediction_column,ind))
  #the notebook may be interruoted due to the lack of memory and can be restarted several times to predict from all models
  if '%s_%s'%(prediction_column,ind) in df.columns:
    print('%s model is already processed'%m)
    continue
  print('Processing %s model'%m)
  Model_foldername = m
  Model_full_foldername = os.path.join(Models, Model_foldername)

  predictor = ktrain.load_predictor(Model_full_foldername)

  prediction=predictor.predict_proba(df[text_column].tolist())

  df_prediction = pd.DataFrame(prediction, columns = predictor.get_classes()) 
  df_prediction=df_prediction[[classname]]
  df_prediction.columns=['%s_%s'%(prediction_column,ind)]

  df['%s_%s'%(prediction_column,ind)]=df_prediction['%s_%s'%(prediction_column,ind)].to_numpy()
  df.to_csv(test_Messages_full_filename, header=True, index=False)
  
  

ts_end = time.time()
print((ts_end - ts_start)/60)

fit_onecycle_0 model is already processed
fit_onecycle_1 model is already processed
fit_onecycle_2 model is already processed
fit_onecycle_3 model is already processed
fit_onecycle_4 model is already processed
Processing fit_onecycle_5 model


Downloading:   0%|          | 0.00/16.0 [00:00<?, ?B/s]

Downloading:   0%|          | 0.00/481 [00:00<?, ?B/s]

Downloading:   0%|          | 0.00/1.68M [00:00<?, ?B/s]

Downloading:   0%|          | 0.00/1.34M [00:00<?, ?B/s]

Downloading:   0%|          | 0.00/150 [00:00<?, ?B/s]

Processing fit_onecycle_6 model
Processing fit_onecycle_7 model
Processing fit_onecycle_8 model
Processing fit_onecycle_9 model
457.1522248109182


## Average by all models/folds

In [ ]:
df[prediction_column]=df[prediction_column_lists].mean(axis=1)

In [ ]:
df.dtypes

Message_Id                int64
Timestamp                object
Topic_1st_Message        object
Parent_Id                 int64
Author_Id                 int64
Topic_Id                  int64
Topic                    object
Chapter_Id                int64
Chapter                  object
original_message         object
message                  object
cnt_childs                int64
cnt_immediate_childs      int64
author                   object
prediction_0            float64
prediction_1            float64
prediction_2            float64
prediction_3            float64
prediction_4            float64
prediction_5            float64
prediction_6            float64
prediction_7            float64
prediction_8            float64
prediction_9            float64
prediction              float64
dtype: object

## Fraction of messages which looks like "fedulya" messages in each author in the dataset

In [6]:
#only authors which score greater then threshold_fedulya_probability(0.9)
author_totals=df.groupby(['Author_Id']).size().reset_index(name='total_counts').sort_values(['total_counts'],ascending=False)
author_predicted=df[df[prediction_column]>threshold_fedulya_probability].groupby(['Author_Id']).size().reset_index(name='predicted_counts').sort_values(['predicted_counts'],ascending=False)
author_predicted=pd.merge(author_totals,author_predicted, left_on='Author_Id', right_on='Author_Id', how='inner')

In [9]:
#to get one author name per Author_Id (then can vary in messages)
Authors=pd.read_csv(Authors_full_filename, error_bad_lines=False, index_col=False)
author_predicted=pd.merge(Authors[['Author_Id','author']],author_predicted, left_on='Author_Id', right_on='Author_Id', how='inner')

In [10]:
#fraction of messages which looks like "fedulya"
author_predicted['fraction_%s_messages'%classname]=author_predicted['predicted_counts']/author_predicted['total_counts']

In [11]:
author_predicted.sort_values(['fraction_%s_messages'%classname],ascending=False, inplace=True)

## fedulya clones

In [12]:
author_predicted[author_predicted['fraction_fedulya_messages']>threshold_frc_fedulya_messages  ]

Author_Id            author  ...  predicted_counts  fraction_fedulya_messages
100     186159         ФедЮля **  ...                95                   0.904762
125     250487          ФедюлЯ +  ...              3795                   0.837748
7        17154  julia4evER **K**  ...                66                   0.750000
140     265057      Julia4eveR _  ...               395                   0.749526
145     266936          ФедЮлЯ _  ...               180                   0.740741
161     282313       Хули-я??? _  ...                81                   0.729730
163     283493        Хули-я?! *  ...              2044                   0.704828

[7 rows x 5 columns]

Let's check the rest of the authors with smaller score

In [13]:
author_predicted.head(30)

Author_Id  ... fraction_fedulya_messages
100     186159  ...                  0.904762
125     250487  ...                  0.837748
7        17154  ...                  0.750000
140     265057  ...                  0.749526
145     266936  ...                  0.740741
161     282313  ...                  0.729730
163     283493  ...                  0.704828
148     269514  ...                  0.675429
155     275933  ...                  0.673567
200     348002  ...                  0.635688
255     482362  ...                  0.300300
252     474446  ...                  0.285714
247     466036  ...                  0.277534
249     472658  ...                  0.271812
261     498092  ...                  0.145228
259     496807  ...                  0.124211
263     499195  ...                  0.111111
262     498614  ...                  0.091727
272     541136  ...                  0.084142
273     541604  ...                  0.049383
25       36280  ...                  0.033784
57       92280  ...                  0.027523
246     464083  ...                  0.025575
129     254769  ...                  0.021583
275     559544  ...                  0.019417
19       30939  ...                  0.019231
211     373392  ...                  0.019231
245     463589  ...                  0.018957
91      151191  ...                  0.017647
127     253790  ...                  0.016461

[30 rows x 5 columns]

Looks like everybody including gutalax _ are clones with a fraction of messages which looks like "fedulya" >=0.6. In fact, they all have nick names related to russian name "Юля" (Julia) in different trabscription.The other group of authors with the fraction ~0.3 have a similar name to ФедЮля ** and might be also clones of the same author. Users from Жгучий перчик + to халоПЕньо + are also clones.

In [26]:
threshold_frc_fedulya_messages=0.04

In [27]:
author_predicted[author_predicted['fraction_fedulya_messages']>threshold_frc_fedulya_messages  ]

Author_Id  ... fraction_fedulya_messages
100     186159  ...                  0.904762
125     250487  ...                  0.837748
7        17154  ...                  0.750000
140     265057  ...                  0.749526
145     266936  ...                  0.740741
161     282313  ...                  0.729730
163     283493  ...                  0.704828
148     269514  ...                  0.675429
155     275933  ...                  0.673567
200     348002  ...                  0.635688
255     482362  ...                  0.300300
252     474446  ...                  0.285714
247     466036  ...                  0.277534
249     472658  ...                  0.271812
261     498092  ...                  0.145228
259     496807  ...                  0.124211
263     499195  ...                  0.111111
262     498614  ...                  0.091727
272     541136  ...                  0.084142
273     541604  ...                  0.049383

[20 rows x 5 columns]

In [ ]:
#df.to_csv(test_Messages_full_filename, header=True, index=False)

### All ФедЮля ** users

In [28]:
fedulya = author_predicted[author_predicted['fraction_fedulya_messages']>threshold_frc_fedulya_messages  ].copy(deep=True)

In [29]:
Authors[((Authors['Author_Id'].isin(fedulya['Author_Id'].tolist())))][['Author_Id','author', 'cnt_messages', 'cnt_topics_started','first_message_ts','last_message_ts']].sort_values('first_message_ts',ascending=True)

Author_Id  ...      last_message_ts
13120     250487  ...  2010-02-22 16:43:00
10417     186159  ...  2010-02-19 03:26:00
13726     265057  ...  2010-03-12 02:26:00
13808     266936  ...  2010-03-23 23:41:00
13926     269514  ...  2010-06-09 21:45:00
14165     275933  ...  2010-08-17 19:19:00
606        17154  ...  2010-08-18 00:12:00
14371     282313  ...  2010-08-27 02:34:00
14419     283493  ...  2010-12-07 01:30:00
15733     348002  ...  2011-12-22 10:24:00
17469     466036  ...  2015-10-06 14:09:00
17565     472658  ...  2015-12-23 23:48:00
17612     474446  ...  2015-12-27 23:30:00
17777     482362  ...  2015-12-26 00:48:00
18122     496807  ...  2017-04-19 20:32:00
18148     498092  ...  2017-01-25 23:27:00
18159     498614  ...  2016-12-15 21:20:00
18169     499195  ...  2017-01-28 03:26:00
18775     541136  ...  2018-01-14 18:30:00
18782     541604  ...  2018-01-17 15:39:00

[20 rows x 6 columns]

The date range when users posted their messages are pretty consistent till some point. Some of date ranges are intersected but it's clear when they were created one by one till a gap 2012 - 2015. If the next users Федюля Иммодиум для анусов ** (466036) - Федюля - ЛопедиуМ для анусов + (482362) are indeed the same person, the gap explains why the fraction of similar messages are low in these users. Something changed in the life of the author as well as her style.

The next group of authors (from жгучий перчик F, 496807 to the last one, халоПЕньо +,	541604) have even less fraction of similar messages. But it looks like they were created even later  after a short gap in 6 months. Maybe belong to the sae user or maybe not.

In [21]:
#authors segmented into clusters - not all discovered authors can be in "CoreAuthors" due to low number of medium sized messages
CoreAuthors = pd.read_csv(CoreAuthors_full_filename, error_bad_lines=False, index_col=False) 

In [30]:
CoreAuthors[CoreAuthors['Author_Id'].isin(fedulya['Author_Id'].tolist())][['Author_Id','author','first_message_ts', 'last_message_ts','cnt_messages','cnt_medium_messages','normalized_kmeans']].sort_values('first_message_ts',ascending=True)

Author_Id  ... normalized_kmeans
1988     250487  ...                 1
1710     186159  ...                 1
2060     265057  ...                 1
2068     266936  ...                 1
2075     269514  ...                 1
2103     275933  ...                 1
2116     282313  ...                 1
2123     283493  ...                 1
2244     348002  ...                 1
2359     466036  ...                 1
2364     472658  ...                 1
2370     474446  ...                 1
2376     482362  ...                 1
2394     496807  ...                 1
2395     498092  ...                 1
2397     498614  ...                 1
2398     499195  ...                 1
2424     541136  ...                 1

[18 rows x 7 columns]

Well, they all belong to the same cluster and might be, indeed, the same person.

In [32]:
print('Total number of fedulya`s posts is %s'%Authors[((Authors['Author_Id'].isin(fedulya['Author_Id'].tolist())))]['cnt_messages'].sum())

Total number of fedulya`s posts is 24990


In [33]:
print('Total number of topics fedulya started is %s'%Authors[((Authors['Author_Id'].isin(fedulya['Author_Id'].tolist())))]['cnt_topics_started'].sum())

Total number of topics fedulya started is 53


They all quite active users with a lot of posts and new topics started. Each has a lot of feedback (responses to their posts).

In [34]:
tmp_df=df[df['Author_Id'].isin(fedulya['Author_Id'].tolist())][['Topic_Id','Author_Id']].groupby(['Topic_Id']).nunique().reset_index().sort_values('Author_Id', ascending=False)
tmp_df.columns=['Topic_Id','cnt_fedulya_authors']
tmp_df=tmp_df[tmp_df['cnt_fedulya_authors']>1]
print('Number of topics where fedulya is talking to herself using different users is %s'%len(tmp_df))

Number of topics where fedulya is talking to herself using different users is 48


*Need full dataset to check if she replies to her own messages. df is only named users (no anonymous) and intermediate replies can be skipped, so the chain of a conversation may have gaps.*

In [35]:
Authors[((Authors['Author_Id'].isin(fedulya['Author_Id'].tolist())))][['Author_Id','author','first_message_ts','fraction_of_messages_with_emojis','fraction_of_messages_with_images','fraction_of_messages_with_links','avg_message_words','avg_original_paragraphs','avg_actual_paragraphs','avg_paragraph_words']].sort_values('first_message_ts',ascending=True)

Author_Id  ... avg_paragraph_words
13120     250487  ...            3.402983
10417     186159  ...            2.946439
13726     265057  ...            4.633349
13808     266936  ...            5.514579
13926     269514  ...            4.102231
14165     275933  ...            3.775667
606        17154  ...            3.007759
14371     282313  ...            3.905285
14419     283493  ...            4.022283
15733     348002  ...            2.525537
17469     466036  ...            3.906661
17565     472658  ...            4.287161
17612     474446  ...            4.438587
17777     482362  ...            3.961199
18122     496807  ...            3.397166
18148     498092  ...            3.968678
18159     498614  ...            3.919372
18169     499195  ...            4.142593
18775     541136  ...            2.893806
18782     541604  ...            2.403153

[20 rows x 10 columns]

*Looks like using emojies makes original, not clean text in the model is more successfull then cleaned, without html tags text. Number of emojis changed at "gutalax _" user. Well, what if starting from "gutalax _" is an other user with clones?* 

In [36]:
Authors[( (Authors['Author_Id'].isin(fedulya['Author_Id'].tolist())))][['Author_Id','author','first_message_ts','cnt_messages','fraction_of_neg_rersponses','fraction_of_pos_rersponses']].sort_values('first_message_ts',ascending=True)

Author_Id  ... fraction_of_pos_rersponses
13120     250487  ...                   0.244973
10417     186159  ...                   0.258333
13726     265057  ...                   0.242947
13808     266936  ...                   0.228374
13926     269514  ...                   0.234098
14165     275933  ...                   0.235901
606        17154  ...                   0.243478
14371     282313  ...                   0.210084
14419     283493  ...                   0.217509
15733     348002  ...                   0.259386
17469     466036  ...                   0.221645
17565     472658  ...                   0.227410
17612     474446  ...                   0.209711
17777     482362  ...                   0.224138
18122     496807  ...                   0.214776
18148     498092  ...                   0.257576
18159     498614  ...                   0.198997
18169     499195  ...                   0.195980
18775     541136  ...                   0.154891
18782     541604  ...                   0.134831

[20 rows x 6 columns]

*Responses to her messages are rather negative then positive. 

In [37]:
Authors[((Authors['Author_Id'].isin(fedulya['Author_Id'].tolist())))][['Author_Id','author','first_message_ts','cnt_messages','fraction_of_tneg_messages','fraction_of_tpos_messages']].sort_values('first_message_ts',ascending=True)

Author_Id  ... fraction_of_tpos_messages
13120     250487  ...                  0.198675
10417     186159  ...                  0.133333
13726     265057  ...                  0.163188
13808     266936  ...                  0.189300
13926     269514  ...                  0.158798
14165     275933  ...                  0.102707
606        17154  ...                  0.090909
14371     282313  ...                  0.099099
14419     283493  ...                  0.107241
15733     348002  ...                  0.219331
17469     466036  ...                  0.163329
17565     472658  ...                  0.127517
17612     474446  ...                  0.164916
17777     482362  ...                  0.147147
18122     496807  ...                  0.133053
18148     498092  ...                  0.141079
18159     498614  ...                  0.124101
18169     499195  ...                  0.202020
18775     541136  ...                  0.174757
18782     541604  ...                  0.160494

[20 rows x 6 columns]

*But her own messages are very negative.*